### DISCUSSION TITLE
Why can't users turn off the admin tracking cookie tk_ai?

### DISCUSSION THREAD

**Discussion by BowFarmer** [woocommerce/woocommerce#49650] [General]
Why is there no option for admin users to turn off the tk_ai tracking cookie? The file includes/tracks/class-wc-tracks-client.php sets this cookie even if admin users don't want to be tracked this way.
The tk_ai tracking cookie serves no function other than to spy on how admin users use Woocommerce.
I try to avoid as many cookies as possible on my site. I don't need another tracking cookie put there by Woocommerce without my permission.

**Comment by senadir** [comment]
Thank you for the report @BowFarmer , this is already reported and fixed here https://github.com/woocommerce/woocommerce/pull/47863 which ships with WooCommerce 9.2

You can disable the cookie being set via this snippet:

```php
add_filter( 'woocommerce_set_cookie_enabled', function( $enabled, $cookie_name ) {
	return'tk_ai' === $cookie_name ? false : $enabled;
}, 10, 2 );
```

As for your concerns.

> The tk_ai tracking cookie serves no function other than to spy on how admin users use Woocommerce.

We use the cookie to anonymously identify a session, this helps us know if a merchant is stuck in some flow, what are the most used flows, and what we should optimize/work on next.

> I don't need another tracking cookie put there by Woocommerce without my permission.

That's a valid concern.

The cookie being set doesn't mean its sending any tracking data, as you need to have the tracking opt in enabled for that.

**Comment by BowFarmer** [comment]
Thank you for the quick answer.
Valid concerns or not, adding cookies to a website without permission or without a way for users to stop them is not ethical. Unethical behavior like this is not acceptable. Just because you can does not mean you should. Programmers need to be ethical. If you want to know if a merchant is stuck in some flow, add some verbiage in the settings explaining what you are doing and how the cookie 'tk_ai' will help you detect this. And let admins decide if they want to participate or not.

**Reply by pjv** [reply]
I concur about the unethical-ness of creating a tracking cookie that is opt-out rather than opt-in.

That current Woo developers would choose to do this is one small indication of what feels like an overall movement in a sad direction for Woo.

A human invariant: the bigger an organization becomes, the worse it becomes. It’s very sad to watch.

**Reply by senadir** [reply]
> Valid concerns or not, adding cookies to a website without permission or without a way for users to stop them is not ethical.

The reason is a bug, not us sneaking to add a cookie that can't be used for anything :D 

> That current Woo developers would choose to do this is one small indication of what feels like an overall movement in a sad direction for Woo.

What's the small indication here? collecting anonymous behavior tracking is a common industry method to measure if something you built is working or not. Those tracking are not used to identify anyone or serve them ads or try to sell them a product.

**Comment by BowFarmer** [comment]
Standard behavior or not, that Woo flippantly adds a cookie is a concern. As a website owner when I see a cookie pop up on my website that I didn't set and I have no idea where it came from, it sets off all sorts of alarm bells. Has my site been hacked? Who is tracking me? Why?
As a developer, Woo needs to get permission from its users before it adds any cookie. The concern is that Woo didn't even consider, how will website owners react if they see a new cookie on their website? This is where Woo needs to instill a sense of ethics in its developers. Just because someone is using your software on their website does not make their website yours. They are using your software because they trust you. Start adding cookies without the users consent and that trust evaporates and users start looking for other solutions. Woo needs to be very careful. It doesn't matter what Woo's intention is. Using cookies is not a trivial thing to do. There are laws governing what website owners can do with cookies so when one pops up like this, it is not a trivial matter.

**Reply by senadir** [reply]
> This is where Woo needs to instill a sense of ethics in its developers.

I think your question that we, Woo developers, are not ethical, is an insulting one to say the least. Especially because I'm engaging you with honesty and apologizing for such bug.

As I've explained, we take those things very seriously. The cookie above was a bug in logic that resulted in setting said cookie, but the cookie in itself is not tracking, it's a single site cookie, and is only set for admins.

The last time [something similar happened](https://github.com/woocommerce/woocommerce/pull/47938), we mobilized most of our engineers to fix this retrospectively in several past releases.

**Reply by BowFarmer** [reply]
I remember looking at discussions at the time between users and Woo, and Woo support staff was saying there was nothing users could do to stop the cookie. The attitude from Woo was flippant. It was like, it's no big deal. I had to write code to stop the cookie. There was no immediate response from Woo at all when this cookie popped up.
